packages/itwinui-react/src/core/ComboBox/ComboBox.test.tsx

```diff
@@ -699,12 +726,21 @@ it('should programmatically clear value', async () => {
   await userEvent.click(screen.getByText('Item 2'));
   const input = container.querySelector('.iui-input') as HTMLInputElement;
   expect(mockOnChange).toHaveBeenCalledWith(2);
-  expect(input).toHaveValue('Item 2');
+
+  // In controlled state, passed value should take priority
+  expect(input).toHaveValue('Item 1');
 
   rerender(
     <ComboBox options={options} onChange={mockOnChange} value={undefined} />,
   );
   expect(mockOnChange).not.toHaveBeenCalledTimes(2);
+
+  // value={undefined} = uncontrolled state. Thus, the value should be the selected value
+  expect(input).toHaveValue('Item 2');
```

Comment:
Can you explain how this works? Passing `value={undefined}` will randomly revert the component to some stale internal state?

Reply:
> Earlier, passing `value={undefined}` = no item selected. But after this PR fixed the controlled state issue, `value={undefined}` = uncontrolled state. After this PR, to force the "no item selected" state, `value={null}` needs to be passed. This is similar to `Select` ([Select code](https://github.com/iTwin/iTwinUI/blob/1232186b654be7769da7477663baf8e7ce239777/packages/itwinui-react/src/core/Select/Select.tsx#L578)).

I think this would be a breaking change. Passing `value={undefined}` should continue to reset the value. We can make it uncontrolled after resetting, but the resetting is important.

Reply:
> Can you explain how this works? Passing `value={undefined}` will randomly revert the component to some stale internal state?

Currently, passing `value={undefined}` reverts back to what is set in the uncontrolled state. Here, the uncontrolled state is `Item 2` since that was clicked earlier on [L726](https://github.com/iTwin/iTwinUI/pull/2048/files/6355ce0e8062070216e558d09a8cf361664c2a71#diff-669ab1d22f5bd28b68a345f18037f496716dea1e95bffba5ff13bf7d6d064b77R726).

I wasn't sure if it was alright that the uncontrolled state can be out of sync with the controlled state. But since `useControlledState` worked that way, I didn't spend too much time thinking about updating that code. But I can investigate this more if the behavior is not ideal.

Reply:
Similar to https://github.com/iTwin/iTwinUI/pull/2048#discussion_r1605191007, you could remove the `useLayoutEffect`?

Reply:
Yes, handled in https://github.com/iTwin/iTwinUI/pull/2048#discussion_r1605315949.